### DISCUSSION TITLE
Get entity with all fields from cache.

### DISCUSSION THREAD

**Discussion by saivanji** [urql-graphql/urql#1110] [Q&A] [answered]
I have the following mutation:
```graphql
renameAction(actionId: UUID!, name: String!): Action!
```

`Action` type is having other different fields apart from `name` and in my optimistic update function I can not be sure what fields will be requested by mutations. So I'm thinking the best way is to return all existing cached `Action` fields + updated name. But the problem is that in order to get them I have to specify all of them manually in `readFragment` function. If I will add new field to schema, I have to update corresponding fragment as well.

Is there a better way to handle such cases with optimistic updates other than specifying manually all entity fields while reading it from the cache?

**Comment by kitten** [comment, marked as answer]
Ultimately you will get warnings if you're missing a field in your optimistic result, along the lines of:

> (13) Invalid undefined: [...] Your optimistic result may be missing a field!

But you may notice that this will be issued as a warning. That's because you may have forgotten to return an update for a specific field, but that doesn't mean that this field is necessarily changing. So it's actually safe to ignore this warning at times (and all warnings are muted in production)

So in theory, all fields that you're reading in `readFragment` to "complete" your optimistic result can be left out. Actually we wanted to ignore this warning if the field was explicitly set to `undefined`, but I believe we've never done that since that still meant that you would have to set all fields to undefined.

Basically, if you're just renaming your action, then it stands to reason that your mutation will only change (and contain) a couple of fields. So given that, this could make your optimistic result very short in turn as well.

**Reply by saivanji** [reply]
@kitten I see, so if I will have the following optimistic update:
```js
export default ({ actionId, name }) => {
  const __typename = "Action"

  return {
    __typename,
    id: actionId,
    name,
  }
}
```

And then execute mutation getting `type`(which was not changed) field from a result it will return it's value without a problem?
```graphql
mutation($actionId: UUID!, $name: String!) {
  renameAction(actionId: $actionId, name: $name) {
    id
    type # that field was not returned from optimistic result
  }
}
```

**Reply by kitten** [reply]
Yea, so basically, an optimistic result isn't meant to be seen in the UI but only used to actually update the internal cache data, so it should be fine. Also if you know that `type` never changes, you could consider to not request it at all. That being said, if you do need it once the mutation resolves, then you can obviously keep it in there, but leaving it out of the optimistic result should be fine.

**Reply by saivanji** [reply]
Great. That is clear now, thanks
